dimod/generators/gates.py

```diff
@@ -0,0 +1,210 @@
+# Copyright 2021 D-Wave Systems Inc.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+
+from dimod.binary.binary_quadratic_model import BinaryQuadraticModel
+from dimod.typing import Variable
+from dimod.vartypes import Vartype
+
+
+__all__ = ['and_gate',
+           'fulladder_gate',
+           'halfadder_gate',
+           'or_gate',
+           'xor_gate',
+           ]
+
+
+def and_gate(in0: Variable, in1: Variable, out: Variable,
+             *, strength: float = 1.0) -> BinaryQuadraticModel:
+    """Return a binary quadratic model with ground states corresponding to an
+    AND gate.
+
+    Args:
+        in0: The variable label for one of the inputs.
+        in1: The variable label for one of the inputs.
+        out: The variable label for the output.
+        strength: The energy of the first infeasible state.
```

Comment:
"first" --> we talk about the "first excited state" and that makes sense but here first is likely to be understood as the first combination of the 8 `xxx` binary states rather then the lowest-energy infeasible state.